File: src/Array/Easy/FindTheHighestAltitude.java
package Array.Easy;

import java.util.Arrays;

public class FindTheHighestAltitude {
    public static void main(String[] args) {
        FindTheHighestAltitude o = new FindTheHighestAltitude();

        System.out.println(
                Arrays.toString(
                        new int[]{
                                o.largestAltitude(new int[]{-5, 1, 5, 0, -7}),
                                o.largestAltitude(new int[]{-4, -3, -2, -1, 4, 3, 2})
                        }
                )
        );
        // [1, 0]
    }

    public int largestAltitude(int[] gain) {
        int[] points = new int[gain.length + 1];
        int max = Integer.MIN_VALUE;

        for (int i = 0; i < gain.length; i++) {
            points[i + 1] = gain[i] + points[i];
            max = Math.max(max, points[i]);
        }

        max = Math.max(max, points[gain.length]);

        return max;
    }
}

// https://leetcode.com/problems/find-the-highest-altitude/

File: src/golang/Array/Easy/find_the_highest_altitude.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(largestAltitude([]int{-5, 1, 5, 0, -7}))
	fmt.Println(largestAltitude([]int{-4, -3, -2, -1, 4, 3, 2}))

	// 1
	// 0
}

func largestAltitude(gain []int) int {
	var points []int = make([]int, len(gain)+1)
	var max int = -int(^uint(0)>>1) - 1

	for i := 0; i < len(gain); i++ {
		points[i+1] = gain[i] + points[i]
		max = maxInt(max, points[i])
	}

	max = maxInt(max, points[len(gain)])

	return max
}

func maxInt(x int, y int) int {
	if x > y {
		return x
	}
	return y
}
